narwhal/consensus/src/bullshark.rs

```diff
@@ -316,13 +275,106 @@ impl Bullshark {
         Ok((Outcome::Commit, committed_sub_dags))
     }
 
+    /// Order the past leaders that we didn't already commit.
+    pub fn order_leaders(&self, leader: &Certificate, state: &ConsensusState) -> Vec<Certificate> {
+        let mut to_commit = vec![leader.clone()];
+        let mut leader = leader;
+        assert_eq!(leader.round() % 2, 0);
+        for r in (state.last_round.committed_round + 2..=leader.round() - 2)
+            .rev()
+            .step_by(2)
+        {
+            // Get the certificate proposed by the previous leader.
+            let (prev_leader, authority) =
+                match self.leader_schedule.leader_certificate(r, &state.dag) {
+                    (authority, Some(x)) => (x, authority),
+                    (authority, None) => {
+                        self.metrics
+                            .leader_election
+                            .with_label_values(&["not_found", authority.hostname()])
+                            .inc();
+
+                        continue;
+                    }
+                };
+
+            // Check whether there is a path between the last two leaders.
+            if self.linked(leader, prev_leader, &state.dag) {
+                to_commit.push(prev_leader.clone());
+                leader = prev_leader;
+            } else {
+                self.metrics
+                    .leader_election
+                    .with_label_values(&["no_path", authority.hostname()])
+                    .inc();
+            }
+        }
+
+        // Now just report all the found leaders
+        let committee = self.committee.clone();
+        let metrics = self.metrics.clone();
+
+        to_commit.iter().for_each(|certificate| {
+            let authority = committee.authority(&certificate.origin()).unwrap();
+
+            metrics
+                .leader_election
+                .with_label_values(&["committed", authority.hostname()])
+                .inc();
+        });
+
+        to_commit
+    }
+
+    /// Checks if there is a path between two leaders.
+    fn linked(&self, leader: &Certificate, prev_leader: &Certificate, dag: &Dag) -> bool {
+        let mut parents = vec![leader];
+        for r in (prev_leader.round()..leader.round()).rev() {
+            parents = dag
+                .get(&r)
+                .expect("We should have the whole history by now")
+                .values()
+                .filter(|(digest, _)| {
+                    parents
+                        .iter()
+                        .any(|x| x.header().parents().contains(digest))
+                })
+                .map(|(_, certificate)| certificate)
+                .collect();
+        }
+        parents.contains(&prev_leader)
+    }
+
+    // When the provided `reputation_scores` are "final" for the current schedule window, then we
```

Comment:
LeaderSchedule is used outside of consensus in proposer, so there seems to be a race: proposer may read leader at round 1002 with old schedule, but after processing round 1000 the schedule is updated. The race in proposer will not hurt correctness, but may hurt liveness. Also it is confusing in general.

Reply:
You are right about the race and that is fine. At some point we can't really avoid this - on the example that you mentioned - as proposer will always (well for the common case) be ahead of consensus and there is no way for us to know , unless we wait for consensus to commit before we proceed , which is practically not possible. Given that we also swap only `f` nodes from the schedule, there is good amount of overlap between schedules, so I estimate that it won't hurt liveness badly, but that's something we have to observe in practice I guess.

Why is it confusing in general? Could you please expand a little bit on that? I mean, given that the leader schedule is now fully dictated by consensus I don't see many different ways of how we could share that info with the proposer, unless we do some static initialisations which I avoided as I didn't consider good practice using global visibility variables for that case.

Reply:
Discussed off-thread. Right now we don't need and don't want the additional complexity to provide proposer with accurate leader info for a given round. But we will add a comment in proposer to make it clear that in some cases leader for a given round from LeaderSchedule may not be the actual leader.

Reply:
happy to clarify anything here as well, in general there is no way to make sure everyone sees the same leader as the commits are done at different rounds in different validators. However since between two consecutive swap tables there is at least f+1 by stake overlap we know that as long as all honest nodes can commit once in 300 rounds we will get liveness